File: docs/revisit.js
/* bookmarker — the arithmetic behind Revisit.
 *
 * "Captured, organised, revisited, shared" — this file is the third verb. It
 * answers, from the library alone: what did you save this week, what do you
 * keep coming back to, what did you save and never look at, what did a month
 * ago look like, which side quests are half-done, and what is shifting.
 *
 * Pure by design — rows in, plain objects out, no DOM, no clock of its own
 * (`now` is a parameter). The Revisit tab renders these objects today; a weekly
 * email digest can build the same ones tomorrow without touching the page.
 *
 * Depends on one thing the page defines, read at call time only so load order
 * doesn't matter: TOPIC_BY_ID (docs/taxonomy.js → index.html). Without it the
 * topic sections fall back to raw ids rather than throwing.
 *
 * Rows are the wire shape (docs/index.html): saved_at, platform, category_id,
 * subcategory, tags, plus the local open ledger the page hangs on them —
 * open_count and last_opened_at, which the sync row has never carried.
 */
"use strict";

const Revisit = (() => {
  const DAY = 86_400_000;

  const savedAt = (r) => Date.parse(r && r.saved_at) || 0;
  const opens = (r) => Math.max(0, Number(r && r.open_count) || 0);
  const openedAt = (r) => Date.parse(r && r.last_opened_at) || 0;
  const byId = (a, b) => (a < b ? -1 : a > b ? 1 : 0);

  /* The taxonomy, if the page loaded it. A topic we can't name is a topic we
     can't link to either, so it stays out of the counts. */
  const topics = () => (typeof TOPIC_BY_ID === "object" && TOPIC_BY_ID) || null;
  const known = (id) => !!id && (!topics() || !!topics()[id]);
  const topicName = (id) => (topics() && topics()[id] && topics()[id].name) || id;

  const between = (rows, from, to) => rows.filter(r => { const t = savedAt(r); return t >= from && t < to; });
  const since = (rows, from) => rows.filter(r => savedAt(r) >= from);

  /** Sources, biggest first: [{ key: "instagram", n: 6 }, …]. The caller names
      them — PLATFORM lives on the page, not here. */
  function platformSplit(rows) {
    const counts = new Map();
    for (const r of rows) {
      const k = (r && r.platform) || "web";
      counts.set(k, (counts.get(k) || 0) + 1);
    }
    return [...counts.entries()]
      .sort((a, b) => b[1] - a[1] || byId(a[0], b[0]))
      .map(([key, n]) => ({ key, n }));
  }

  function topicCounts(rows) {
    const counts = new Map();
    for (const r of rows) if (r && known(r.category_id)) counts.set(r.category_id, (counts.get(r.category_id) || 0) + 1);
    return counts;
  }

  /** Topics that grew, this window against the one before it. */
  function risingTopics(now, before, limit = 3) {
    const a = topicCounts(now), b = topicCounts(before);
    const out = [];
    for (const [id, n] of a) {
      const prev = b.get(id) || 0;
      if (n <= prev) continue;
      out.push({ id, name: topicName(id), n, prev, delta: n - prev });
    }
    return out.sort((x, y) => y.delta - x.delta || y.n - x.n || byId(x.id, y.id)).slice(0, limit);
  }

  /* A shift is a change in *share*, not in count — saving more of everything
     isn't a shift. Small wobbles and one-off topics aren't either, so a move
     has to clear three points and rest on more than a single bork. */
  const SHIFT_MIN_DELTA = 0.03;
  const SHIFT_MIN_BORKS = 2;

  function shifting(now, before, limit = 3) {
    if (!now.length || !before.length) return null;   // no "before" is not a trend
    const a = topicCounts(now), b = topicCounts(before);
    const items = [];
    for (const id of new Set([...a.keys(), ...b.keys()])) {
      const n = a.get(id) || 0, prev = b.get(id) || 0;
      if (Math.max(n, prev) < SHIFT_MIN_BORKS) continue;
      const share = n / now.length, prevShare = prev / before.length;
      const delta = share - prevShare;
      if (Math.abs(delta) < SHIFT_MIN_DELTA) continue;
      items.push({ id, name: topicName(id), n, prev, share, prevShare, delta, direction: delta > 0 ? "up" : "down" });
    }
    const top = items
      .sort((x, y) => Math.abs(y.delta) - Math.abs(x.delta) || byId(x.id, y.id))
      .slice(0, limit);
    if (!top.length) return null;
    return { items: top, line: shiftLine(top) };
  }

  /** "More Fitness, less Crypto" — the sentence, from the parts. */
  function shiftLine(items) {
    const line = items.map(i => `${i.direction === "up" ? "more" : "less"} ${i.name}`).join(", ");
    return line.charAt(0).toUpperCase() + line.slice(1);
  }

  const NEVER_SHOWN = 12;

  /**
   * @param rows     every live bork (deleted rows are dropped here anyway)
   * @param missions side quests, already filtered to the live ones
   * @param now      epoch ms — the only clock this file has
   * @returns { total, thin, week, repeat, never, monthAgo, quests, shifting }
   *          with every section either a plain object or null when it's empty.
   */
  function build(rows, missions, now) {
    const all = (Array.isArray(rows) ? rows : []).filter(r => r && !r.deleted_at);
    const t = Number(now) || Date.now();
    const total = all.length;

    // 1 · Saved this week, against the week before it.
    const thisWeek = since(all, t - 7 * DAY);
    const lastWeek = between(all, t - 14 * DAY, t - 7 * DAY);
    const week = thisWeek.length
      ? { n: thisWeek.length, platforms: platformSplit(thisWeek), rising: risingTopics(thisWeek, lastWeek) }
      : null;

    // 2 · What you actually go back to. Ties break on the most recent open.
    const reopened = all.filter(r => opens(r) > 0)
      .sort((a, b) => opens(b) - opens(a) || openedAt(b) - openedAt(a) || savedAt(b) - savedAt(a));
    const repeat = reopened.length ? { rows: reopened.slice(0, 5) } : null;

    // 3 · The pile. Saved long enough ago to count as ignored, never opened.
    const forgotten = all.filter(r => opens(r) === 0 && savedAt(r) > 0 && savedAt(r) < t - 7 * DAY)
      .sort((a, b) => savedAt(b) - savedAt(a));
    const never = forgotten.length
      ? { rows: forgotten, n: forgotten.length, shown: Math.min(NEVER_SHOWN, forgotten.length),
          more: Math.max(0, forgotten.length - NEVER_SHOWN) }
      : null;

    // 4 · A month ago today.
    const monthRows = between(all, t - 35 * DAY, t - 28 * DAY).sort((a, b) => savedAt(b) - savedAt(a)).slice(0, 3);
    const monthAgo = monthRows.length ? { rows: monthRows } : null;

    // 5 · Side quests with something left to tick off.
    const open = (Array.isArray(missions) ? missions : [])
      .filter(m => m && !m.deleted_at && !m.is_archived && (m.todos || []).some(td => td && !td.done));
    const quests = open.length ? { missions: open } : null;

    // 6 · What's shifting: the last 30 days against the 30 before, and only
    //     once there's enough on both sides of the line to mean anything.
    const now30 = since(all, t - 30 * DAY);
    const prev30 = between(all, t - 60 * DAY, t - 30 * DAY);
    const shift = now30.length + prev30.length >= 20 ? shifting(now30, prev30) : null;

    return { total, thin: total < 10, week, repeat, never, monthAgo, quests, shifting: shift };
  }

  return { build, DAY, NEVER_SHOWN, SHIFT_MIN_DELTA, SHIFT_MIN_BORKS,
           _platformSplit: platformSplit, _rising: risingTopics, _shifting: shifting, _line: shiftLine };
})();

const buildRevisit = (rows, missions, now) => Revisit.build(rows, missions, now);

/* Node (Scripts/test_revisit.mjs) rather than a browser. */
if (typeof module === "object" && module.exports) module.exports = { Revisit, buildRevisit };

File: Core/Revisit.swift
import Foundation

/// One bork, reduced to what "what should I look at again?" needs.
///
/// Plain values rather than the SwiftData model on purpose — see `Revisit`.
struct RevisitBork: Identifiable, Hashable, Sendable {
    var id: String
    var title: String
    var platform: Platform
    var topicID: String?
    var topicName: String?
    var savedAt: Date
    var openCount: Int
    var lastOpenedAt: Date?
}

/// A side quest with steps left on it.
struct RevisitQuest: Identifiable, Hashable, Sendable {
    var id: String
    var title: String
    var remaining: Int
}

/// What to look at again.
///
/// The library answers "what have I got". Browse answers "where is it". Neither
/// answers the question the product is actually for: *everything interesting
/// you scroll past, captured, organised, **revisited**, shared*. Saving is
/// solved; revisiting was, until now, entirely on the user to remember to do.
///
/// **Why this is a pure type and not a view.** Every section here is a
/// perfectly good paragraph of a weekly email — "you saved 14 things, 11 of
/// them you never opened, a month ago you were reading about X". Computing it
/// inside a SwiftUI view would mean writing it all again the day the digest
/// ships. So the screen renders a value, and the value can be built anywhere:
/// from SwiftData on the phone today, from a Postgres row server-side later.
/// `RevisitBork` deliberately carries the topic *name*, not just its id, so the
/// computation never needs the taxonomy either.
///
/// Sections are ordered by how likely they are to send someone back into their
/// library, not by how clever they are. "Saved, never opened" is the one that
/// earns the tab.
enum Revisit {

    /// Below this the screen has nothing honest to say, and says so.
    static let thinFloor = 10

    /// A bork has to be a week old before "never opened" means anything —
    /// saving something and not opening it that afternoon is normal.
    static let neverOpenedAge = 7
    static let neverOpenedCap = 12

    /// "A month ago" is a window, not a day: an exact 30-day-ago lookup is
    /// empty most days, which makes the section feel broken rather than quiet.
    static let monthAgoRange = 28...35
    static let monthAgoCap = 3

    static let comingBackCap = 5

    /// "What's shifting" needs enough data to not be noise.
    static let shiftFloor = 20
    static let shiftWindow = 30
    static let shiftCap = 3
    /// Share changes smaller than this are rounding, not a trend.
    static let shiftThreshold = 0.03

    // MARK: - Value types

    struct PlatformCount: Hashable, Sendable {
        var name: String
        var count: Int
    }

    struct TopicGrowth: Identifiable, Hashable, Sendable {
        var id: String
        var name: String
        var count: Int
        var previous: Int
        var gain: Int { count - previous }
    }

    struct ThisWeek: Hashable, Sendable {
        var count: Int
        var platforms: [PlatformCount]
        var rising: [TopicGrowth]
        /// "14 borks · 6 Instagram · 5 X · 3 YouTube"
        var headline: String
    }

    struct NeverOpened: Hashable, Sendable {
        /// **All** of them, newest first. The section shows `preview` and
        /// offers the rest behind "and N more", but the model holds the whole
        /// pile: the list screen needs it, and a digest would want the count
        /// rather than "twelve or more".
        var items: [RevisitBork]
        /// "11 things you saved and never looked at."
        var headline: String

        var total: Int { items.count }
        var preview: [RevisitBork] { Array(items.prefix(Revisit.neverOpenedCap)) }
        var overflow: Int { max(0, items.count - Revisit.neverOpenedCap) }
    }

    struct Shift: Identifiable, Hashable, Sendable {
        var id: String
        var name: String
        var share: Double
        var previousShare: Double
        var isUp: Bool
        var magnitude: Double { abs(share - previousShare) }
    }

    struct Model: Hashable, Sendable {
        var total: Int
        /// Under `thinFloor` borks: one warm card instead of six thin sections.
        var isThin: Bool
        var thisWeek: ThisWeek?
        var comingBack: [RevisitBork]
        var neverOpened: NeverOpened?
        var monthAgo: [RevisitBork]
        var quests: [RevisitQuest]
        var shifts: [Shift]
        /// "More Fitness, less Crypto"
        var shiftSentence: String?

        /// True when every section came back empty — a library that is big
        /// enough to have sections but has nothing due.
        var hasNothing: Bool {
            thisWeek == nil && comingBack.isEmpty && neverOpened == nil
                && monthAgo.isEmpty && quests.isEmpty && shifts.isEmpty
        }

        static let empty = Model(
            total: 0, isThin: true, thisWeek: nil, comingBack: [],
            neverOpened: nil, monthAgo: [], quests: [], shifts: [], shiftSentence: nil
        )
    }

    // MARK: - Build

    static func build(
        borks: [RevisitBork],
        quests: [RevisitQuest],
        now: Date = .now,
        calendar: Calendar = .current
    ) -> Model {
        let today = calendar.startOfDay(for: now)
        func daysBack(_ days: Int) -> Date {
            calendar.date(byAdding: .day, value: -days, to: today) ?? today
        }

        let total = borks.count
        let shifts = shifts(borks, now: now, calendar: calendar)

        return Model(
            total: total,
            isThin: total < thinFloor,
            thisWeek: thisWeek(borks, since: daysBack(7), previousFrom: daysBack(14)),
            comingBack: comingBack(borks),
            neverOpened: neverOpened(borks, before: daysBack(neverOpenedAge)),
            // Whole calendar days, inclusive at both ends: the window is
            // "day 28 through day 35", so it ends at the start of day 27.
            monthAgo: monthAgo(
                borks,
                from: daysBack(monthAgoRange.upperBound),
                to: daysBack(monthAgoRange.lowerBound - 1)
            ),
            quests: quests.filter { $0.remaining > 0 },
            shifts: shifts,
            shiftSentence: sentence(for: shifts)
        )
    }

    // MARK: - Sections

    /// 1. Saved this week — the count, where it came from, and which topics
    ///    grew against the week before.
    private static func thisWeek(
        _ borks: [RevisitBork], since: Date, previousFrom: Date
    ) -> ThisWeek? {
        let recent = borks.filter { $0.savedAt >= since }
        guard !recent.isEmpty else { return nil }

        let previous = borks.filter { $0.savedAt >= previousFrom && $0.savedAt < since }

        let platforms = Dictionary(grouping: recent, by: { $0.platform })
            .map { PlatformCount(name: $0.key.name, count: $0.value.count) }
            .sorted { ($0.count, $1.name) > ($1.count, $0.name) }

        var counts: [String: (name: String, now: Int, before: Int)] = [:]
        for bork in recent {
            guard let id = bork.topicID, let name = bork.topicName else { continue }
            counts[id, default: (name, 0, 0)].now += 1
        }
        for bork in previous {
            guard let id = bork.topicID, let name = bork.topicName else { continue }
            counts[id, default: (name, 0, 0)].before += 1
        }

        let rising = counts
            .map { TopicGrowth(id: $0.key, name: $0.value.name, count: $0.value.now, previous: $0.value.before) }
            .filter { $0.gain > 0 }
            .sorted { ($0.gain, $0.count, $1.name) > ($1.gain, $1.count, $0.name) }
            .prefix(3)

        // Four sources is as much as one line can carry legibly; the count is
        // always the whole week regardless.
        let split = platforms.prefix(4).map { "\($0.count) \($0.name)" }
        let headline = ([Copy.countedBorks(recent.count)] + split).joined(separator: " · ")

        return ThisWeek(count: recent.count, platforms: platforms, rising: Array(rising), headline: headline)
    }

    /// 2. You keep coming back to — the honest measure of what mattered.
    private static func comingBack(_ borks: [RevisitBork]) -> [RevisitBork] {
        borks
            .filter { $0.openCount > 0 }
            .sorted { a, b in
                if a.openCount != b.openCount { return a.openCount > b.openCount }
                let left = a.lastOpenedAt ?? .distantPast
                let right = b.lastOpenedAt ?? .distantPast
                if left != right { return left > right }
                return a.id < b.id
            }
            .prefix(comingBackCap)
            .map { $0 }
    }

    /// 3. Saved, never opened. The reason the tab exists: this is the pile
    ///    every other bookmarking app quietly lets you build and never mentions.
    private static func neverOpened(_ borks: [RevisitBork], before cutoff: Date) -> NeverOpened? {
        let stale = borks
            .filter { $0.openCount == 0 && $0.savedAt < cutoff }
            .sorted { $0.savedAt > $1.savedAt }
        guard !stale.isEmpty else { return nil }

        let noun = stale.count == 1 ? "thing" : "things"
        return NeverOpened(
            items: stale,
            headline: "\(stale.count) \(noun) you saved and never looked at."
        )
    }

    /// 4. A month ago today.
    private static func monthAgo(_ borks: [RevisitBork], from: Date, to: Date) -> [RevisitBork] {
        borks
            .filter { $0.savedAt >= from && $0.savedAt < to }
            .sorted { $0.savedAt > $1.savedAt }
            .prefix(monthAgoCap)
            .map { $0 }
    }

    /// 6. What's shifting — share of saves in the last 30 days against the 30
    ///    before. Share rather than raw count, so a quiet month doesn't read as
    ///    "less of everything".
    private static func shifts(_ borks: [RevisitBork], now: Date, calendar: Calendar) -> [Shift] {
        let today = calendar.startOfDay(for: now)
        guard let recentStart = calendar.date(byAdding: .day, value: -shiftWindow, to: today),
              let earlierStart = calendar.date(byAdding: .day, value: -shiftWindow * 2, to: today)
        else { return [] }

        let recent = borks.filter { $0.savedAt >= recentStart }
        let earlier = borks.filter { $0.savedAt >= earlierStart && $0.savedAt < recentStart }
        guard recent.count + earlier.count >= shiftFloor,
              !recent.isEmpty, !earlier.isEmpty else { return [] }

        var names: [String: String] = [:]
        func share(_ items: [RevisitBork]) -> [String: Double] {
            var counts: [String: Int] = [:]
            for item in items {
                guard let id = item.topicID, let name = item.topicName else { continue }
                counts[id, default: 0] += 1
                names[id] = name
            }
            let filed = counts.values.reduce(0, +)
            guard filed > 0 else { return [:] }
            return counts.mapValues { Double($0) / Double(filed) }
        }

        let nowShare = share(recent), beforeShare = share(earlier)

        return Set(nowShare.keys).union(beforeShare.keys)
            .compactMap { id -> Shift? in
                guard let name = names[id] else { return nil }
                let a = nowShare[id] ?? 0, b = beforeShare[id] ?? 0
                guard abs(a - b) >= shiftThreshold else { return nil }
                return Shift(id: id, name: name, share: a, previousShare: b, isUp: a > b)
            }
            .sorted { ($0.magnitude, $1.id) > ($1.magnitude, $0.id) }
            .prefix(shiftCap)
            // Which three is a question of size; what order to *read* them in
            // is a question of copy. "More Fitness, less Crypto" is a better
            // sentence than "Less Crypto, more Fitness", and a headline that
            // opens with what you are doing more of is the one worth reading.
            // Floating-point ties between an equal rise and fall would
            // otherwise decide it, which is no way to pick a sentence.
            .sorted { a, b in
                if a.isUp != b.isUp { return a.isUp }
                if a.magnitude != b.magnitude { return a.magnitude > b.magnitude }
                return a.id < b.id
            }
    }

    /// "More Fitness, less Crypto" — the whole section in one line.
    private static func sentence(for shifts: [Shift]) -> String? {
        guard !shifts.isEmpty else { return nil }
        let parts = shifts.map { $0.isUp ? "more \($0.name)" : "less \($0.name)" }
        let line = parts.joined(separator: ", ")
        return line.prefix(1).uppercased() + line.dropFirst()
    }
}
